pkg/log/log.go

```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2018 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package log provides a logger that satisfies https://github.com/go-logr/logr.
+// It is implemented as a light wrapper around sigs.k8s.io/controller-runtime/pkg/log
+package log
```

Comment:
What is the purpose of this package?

Reply:
It's just somewhere for our global logger to live.

We use our own global logger rather than controller-runtime's instance because if we use controller-runtime's global logger (and supply it with a real underlying logger implementation rather than its default `NullLogger{}`) we get all of the (in my opinion) extremely verbose and questionably useful output of that logger.

I included some `--debug` mode logging output in a previous comment on this PR. All of the log lines that do not mention `crossplane` are coming from the controller-runtime's global logger, at Info level (i.e. `log.Info()` not `log.V(1).Info` per https://github.com/go-logr/zapr#implementation-details).

Reply:
What is the recommendation for crossplane components to handle different log levels, i.e info debug, etc?

Reply:
We need to decide and document these recommendations. I personally would encourage us to use the following pattern per https://dave.cheney.net/2015/11/05/lets-talk-about-logging:

1. Things that developers care about when they are developing or debugging Crossplane. These are logged using `log.V(1).Info()`
1. Things that users care about when using Crossplane. These are logged using `log.Info()`.

https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L235 - this comment makes a good point. We could declare a constant `const DEBUG = 1` so that `log.V(1).Info()` becomes `log.V(logging.DEBUG).Info()`. Seeing debug and info in the same statement is confusing, but it's a _little_ clearer what that the caller is aiming for a debug log than `V(1)` alone is.

Again, I think this is a terrible logging API, but if we want to use controller-runtime's logging patterns and want to have more than just Info log level, this is our best option.

Reply:
Did you also want to incorporate the const for DEBUG = 1 into this PR?

Reply:
Done.